packages/rest/src/http/metadata.ts

```diff
@@ -0,0 +1,436 @@
+import {
+  compilerAssert,
+  DiagnosticCollector,
+  isVisible as isVisibleCore,
+  Model,
+  ModelProperty,
+  Program,
+  Queue,
+  TwoLevelMap,
+  Type,
+  Union,
+  walkPropertiesInherited,
+} from "@cadl-lang/compiler";
+import { isBody, isHeader, isPathParam, isQueryParam, isStatusCode } from "./decorators.js";
+import { HttpVerb } from "./types.js";
+
+/**
+ * Flags enum representation of well-known visibilities that are used in
+ * REST API.
+ */
+export enum Visibility {
+  Read = 1 << 0,
+  Create = 1 << 1,
+  Update = 1 << 2,
+  Delete = 1 << 3,
+  Query = 1 << 4,
+
+  All = Read | Create | Update | Delete | Query,
+
+  /**
+   * Additional flag to indicate when something is nested in a collection
+   * and therefore no metadata is applicable.
+   */
+  Item = 1 << 20,
+}
+
+const visibilityToArrayMap: Map<Visibility, string[]> = new Map();
+function visibilityToArray(visibility: Visibility): readonly string[] {
+  // Item flag is not a real visibility.
+  visibility &= ~Visibility.Item;
+
+  let result = visibilityToArrayMap.get(visibility);
+  if (!result) {
+    result = [];
+
+    if (visibility & Visibility.Read) {
+      result.push("read");
+    }
+    if (visibility & Visibility.Create) {
+      result.push("create");
+    }
+    if (visibility & Visibility.Update) {
+      result.push("update");
+    }
+    if (visibility & Visibility.Delete) {
+      result.push("delete");
+    }
+    if (visibility & Visibility.Query) {
+      result.push("query");
+    }
+
+    compilerAssert(result.length > 0, "invalid visibility");
+    visibilityToArrayMap.set(visibility, result);
+  }
+
+  return result;
+}
+
+/**
+ * Provides a naming suffix to create a unique name for a type with this
+ * visibility.
+ *
+ * `Visibility.All` gets empty suffix, otherwise visibilities are joined in
+ * pascal-case with `Or`. And `Item` is if `Visibility.Item` is produced.
+ *
+ * Examples:
+ *  - Visibility.All => ""
+ *  - Visibility.Read => "Read"
+ *  - Visibility.Create | Visibility.Update => "CreateOrUpdate"
+ *  - Visibility.Create | Visibility.Item => "CreateItem"
+ *  - Visibility.Create | Visibility.Update | Visibility.Item =>  "CreateOrUpdateItem"
+ *  */
+export function getVisibilitySuffix(visibility: Visibility) {
```

Comment:
Maybe @johanste can help us run this by the rest of the API Stewardship Board to make sure they're cool with the automatic schema suffixing and the names we use

Reply:
We should stress that the suffixes are for OpenAPI emit and that a benefit of this approach is that you can more easily directly see the Cadl type if you are emitting Cadl -> not openapi because the author can use the logical type directly.

(*) We still don't handle update -> deeply make optional, though. But perhaps we should? It feels wrong to bind visibility("update") to this, but also maybe it pragmatically closes the last place where you can't just put the logical type?